text/trace/0234-payload-collection.md

```diff
@@ -0,0 +1,363 @@
+  For example - `json`, `protobuf`, `avro`, `utf-8`.
+
+Adding each attribute should not be dependent on the others in any way.
+
+### API
+
+We propose specifying APIs the will abstract the handling and capturing of
+payload data, to allow customizing this functionality and evolving it over time.
+
+This functionality includes:
+
+- Adding attributes with enforced semantics
+- Enforcing limits - e.g. shortening long strings
+- Applying central configuration over the general functionality
+- Parsing raw payload buffers - converting payload data from bytes to OTel attributes, for
+  supported encodings, in an extendable fashion.
+- Aggregating multiple data chunks (to assist with handling asynchronous buffers)
+
+The APIs given here are only a draft to describe general characteristics.
+Like in the semantics, a further discussion will be required before creating actual specifications.
+
+#### `Payload` class
+
+We propose adding a class that will provide the related functionality for handling payload data.
+Both traces and logs could use instances of this class, which will assist sharing related functionality.
+
+For simplicity, we will show a basic example of this class in Python. In reality, an API and its implementation should be separated.
+
+```python
+class Payload:
+    def __init__(self, content: Any, encoding=None, size=None):
+        self.content = self._parse_payload_content(content)
+        self.encoding = encoding
+        self.size = size
+
+    @staticmethod
+    def _parse_payload_content(content: Any) -> Any:
+        # Potentially altering of the content could be added, such as
+        #  format conversion, shortening, filtering, etc.
+        return content
+
+    @classmethod
+    def from_binary(
+        cls,
+        payload_buffer: bytes,
+        encoding: str = 'utf-8',
+    ) -> Payload:
+        if encoding == 'utf-8':
+            content = payload_buffer.decode('utf-8')
+        elif encoding == 'utf-16':
+            content = payload_buffer.decode('utf-16')
+        elif encoding == 'json':
+            content = json.loads(payload_buffer.decode('utf-8'))
+        # ... other supported encodings
+        else:
+            raise ValueError("Unsupported payload encoding %s", encoding)
+
+        return cls(content, encoding=encoding, size=len(payload_buffer))
+```
+
+### Payload attributes in Traces
+
+In order to efficiently collect payload data in traces, this proposal depends on the [support of maps and heterogeneous arrays in spans
+attributes](https://github.com/open-telemetry/opentelemetry-specification/pull/2888).
+
+For adding payload attributes, we propose adding a dedicated method for `Span` class:
+
+```python
+
+def add_payload(self, key: str, payload: Payload):
+  self.set_attributes(
+    {
+      f"{key}.content": payload.content,
+      f"{key}.encoding": payload.encoding,
+      f"{key}.size": payload.size
+    }
+  )
+```
+
+### Payload attributes in Logs
```

Comment:
+1 to resolving this discussion as a priority.

Reply:
But should it affect the attribute names for spans as proposed here?